### PR TITLE
[WIP] Cscli revamp

### PR DESCRIPTION
**Pull request by erenJag**

### CODE REVIEW

File: cmd/crowdsec-cli/decisions.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+			apiclient.BaseURL, err = url.Parse(csConfig.LapiClient.Credentials.Url)
+			if err != nil {
+				log.Fatalf("failed to parse Local API URL %s : %v ", csConfig.LapiClient.Credentials.Url, err.Error())
+			}
+
+			password := strfmt.Password(csConfig.LapiClient.Credentials.Password)
+			t := &apiclient.JWTTransport{
+				MachineID: &csConfig.LapiClient.Credentials.Login,
+				Password:  &password,
+				Scenarios: []string{"aaaaaa", "bbbbb"},
```
Comment (buixor) [inline cmd/crowdsec-cli/decisions.go:101]: I think we already have a method to get the existing scenarios from hub easily

File: cmd/crowdsec-cli/decisions.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+			}
+
+			Client = apiclient.NewClient(t.Client())
+		},
+	}
+
+	var cmdDecisionsList = &cobra.Command{
+		Use:     "list --scope [scope] --value [value] --type [type]",
+		Short:   "List decisions",
+		Long:    `List decisions from the LAPI`,
```
Comment (buixor) [inline cmd/crowdsec-cli/decisions.go:111]: maybe we want something more user focused like "List active decisions" (for the help). If 90% of the users are running single machine setup, we might want to "hide" the complexity of the local API service ?

File: cmd/crowdsec-cli/decisions.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+		Example: `cscli decisions list --scope ip --value 1.2.3.4 --type ban"`,
+		Args:    cobra.MaximumNArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			var err error
+			activeDecision := true
+
+			filter := apiclient.AlertsListOpts{}
```
Comment (buixor) [inline cmd/crowdsec-cli/decisions.go:118]: I think we might want to offer a native filter for IPs, after all this is what most users are after.
Having `--ip x.x.x.x` and `--range x.x.x.x/y` makes sense ?

File: cmd/crowdsec-cli/decisions.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+				log.Fatalf(err.Error())
+			}
+
+			log.Info("Decision successfully added")
+		},
+	}
+	cmdDecisions.AddCommand(cmdDecisionsAdd)
+
+	var cmdDecisionsDelete = &cobra.Command{
+		Use:   "delete (--scope [scope] --value [value] --type [type] | --id [decision_id] | --all)",
```
Comment (buixor) [inline cmd/crowdsec-cli/decisions.go:245]: same as for the `add` command, adding shortcuts for `--ip` and `--range` will make user's life better :)

File: cmd/crowdsec-cli/keys.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+				}
+				table.Render()
+			} else if csConfig.Cscli.Output == "json" {
+				x, err := json.MarshalIndent(blockers, "", " ")
+				if err != nil {
+					log.Fatalf("failed to unmarshal")
+				}
+				fmt.Printf("%s", string(x))
+			} else if csConfig.Cscli.Output == "raw" {
+				for _, b := range blockers {
```
Comment (buixor) [inline cmd/crowdsec-cli/keys.go:79]: beware, here you separate "raw" output with spaces, but it was with `,` in the `ban list` one. I think you want as well to display the "column" names.

File: cmd/crowdsec-cli/keys.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+		},
+	}
+	cmdKeys.AddCommand(cmdKeysList)
+
+	var cmdKeysAdd = &cobra.Command{
+		Use:     "add",
+		Short:   "add api keys",
+		Long:    `add `,
+		Example: `cscli keys add --name test [--ip 1.2.3.4]`,
```
Comment (buixor) [inline cmd/crowdsec-cli/keys.go:100]: does the ip part allows to restrict API key to one source IP ?

File: cmd/crowdsec-cli/keys.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	cmdKeysAdd.Flags().StringVarP(&keyName, "name", "n", "", "name to assigned for the api key")
+	cmdKeysAdd.Flags().StringVarP(&keyIP, "ip", "i", "", "ip address of the blocker")
+	cmdKeysAdd.Flags().IntVarP(&keyLength, "length", "l", 16, "length of the api key")
+	cmdKeys.AddCommand(cmdKeysAdd)
+
+	var cmdKeysDelete = &cobra.Command{
+		Use:     "delete",
+		Short:   "delete api keys",
+		Long:    `delete `,
+		Example: `cscli keys delete --name test [--ip 1.2.3.4]`,
```
Comment (buixor) [inline cmd/crowdsec-cli/keys.go:131]: can't we delete by the key itself ?

File: cmd/crowdsec-cli/main.go [hunk 8]
```diff
@@ -102,7 +105,7 @@ API interaction:
 	}
 	rootCmd.AddCommand(cmdVersion)
 
-	rootCmd.PersistentFlags().StringVarP(&ConfigFilePath, "config", "c", "/etc/crowdsec/config/default.yaml", "path to crowdsec config file")
+	rootCmd.PersistentFlags().StringVarP(&ConfigFilePath, "config", "c", "../../config/dev.yaml", "path to crowdsec config file")
```
Comment (buixor) [inline cmd/crowdsec-cli/main.go:116]: I don't think we want this default :)

File: cmd/crowdsec-cli/watchers.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+var machinePassword string
+var machineIP string
+var interactive bool
+
+func NewWatchersCmd() *cobra.Command {
+	/* ---- DECISIONS COMMAND */
+	var cmdWatchers = &cobra.Command{
+		Use:   "watchers [action]",
+		Short: "Manage local API watchers",
+		Long: `
```
Comment (buixor) [inline cmd/crowdsec-cli/watchers.go:27]: maybe in the "long" description we should clearly differentiate that the API keys are read-only while those can write ? as we're re-using the term "API" I'm afraid we will confuse users.

File: cmd/crowdsec-cli/watchers.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+				log.Errorf("unable to list blockers: %s", err)
+			}
+			if csConfig.Cscli.Output == "human" {
+				table := tablewriter.NewWriter(os.Stdout)
+				table.SetCenterSeparator("")
+				table.SetColumnSeparator("")
+
+				table.SetHeaderAlignment(tablewriter.ALIGN_LEFT)
+				table.SetAlignment(tablewriter.ALIGN_LEFT)
+				table.SetHeader([]string{"Name", "IP Address", "Status"})
```
Comment (buixor) [inline cmd/crowdsec-cli/watchers.go:61]: should we add a "last signal push" or such

File: cmd/crowdsec-cli/watchers.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+				fmt.Printf("%s", string(x))
+			} else if csConfig.Cscli.Output == "raw" {
+				for _, w := range watchers {
+					var validated string
+					if w.IsValidated {
+						validated = "true"
+					} else {
+						validated = "false"
+					}
+					fmt.Printf("%s %s %s\n", w.MachineId, w.IpAddress, validated)
```
Comment (buixor) [inline cmd/crowdsec-cli/watchers.go:86]: I think you want column names here. As for `ban list`, it's not consistent (" " vs ",")

File: cmd/crowdsec-cli/watchers.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+				log.Errorf("unable to create machine: %s", err)
+				return
+			}
+			log.Infof("Machine '%s' created successfully", machineID)
+		},
+	}
+	cmdWatchersAdd.Flags().StringVarP(&machineID, "machine", "m", "", "machine ID to login to the API")
+	cmdWatchersAdd.Flags().StringVarP(&machinePassword, "password", "p", "", "machine password to login to the API")
+	cmdWatchersAdd.Flags().StringVar(&machineIP, "ip", "", "machine ip address")
+	cmdWatchersAdd.Flags().BoolVarP(&interactive, "interactive", "i", false, "machine ip address")
```
Comment (buixor) [inline cmd/crowdsec-cli/watchers.go:127]: flag description is wrong

File: cmd/crowdsec-cli/watchers.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+	cmdWatchers.AddCommand(cmdWatchersAdd)
+
+	var cmdWatchersDelete = &cobra.Command{
+		Use:     "delete",
+		Short:   "delete watchers",
+		Long:    `delete `,
+		Example: `cscli watchers delete --machine test`,
```
Comment (buixor) [inline cmd/crowdsec-cli/watchers.go:134]: should we allow by IP ?

**Review by buixor — CHANGES_REQUESTED** [review]
discussions on WIP

File: cmd/crowdsec-cli/decisions.go (hunk 1 above)
Comment (registergoofy) [reply cmd/crowdsec-cli/decisions.go]: Not in this branch either. This has to be merged

**Review by registergoofy — APPROVED** [review]

**Pull request merged**